keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/constraints/Ec2CanaryConstraintDeployHandler.kt

```diff
@@ -0,0 +1,166 @@
+import com.netflix.spinnaker.keel.retrofit.isNotFound
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import retrofit2.HttpException
+
+class Ec2CanaryConstraintDeployHandler(
+  private val taskLauncher: TaskLauncher,
+  private val cloudDriverService: CloudDriverService,
+  private val cloudDriverCache: CloudDriverCache,
+  private val imageService: ImageService
+) : CanaryConstraintDeployHandler {
+  companion object {
+    const val DEFAULT_KAYENTA_STORAGE_ACCOUNT = "s3-objects"
+    private val log by lazy { LoggerFactory.getLogger(Ec2CanaryConstraintDeployHandler::class.java) }
+  }
+
+  // TODO: The rest of Spinnaker should be refactored to refer to ec2 as ec2 instead of aws.
+  //  But for now, `CanaryConstraint.source` points to a source cluster (including cloud
+  //  provider) as Orca and Clouddriver would - that is, with "aws" as the provider when ec2
+  //  is intended.
+  override val supportedClouds = setOf("ec2", "aws")
+
+  override fun deployCanary(
+    constraint: CanaryConstraint,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    targetEnvironment: Environment,
+    regions: Set<String>
+  ): Map<String, Task> {
+
+    val judge = "canary:${deliveryConfig.application}:${targetEnvironment.name}:${constraint.canaryConfigId}"
+
+    val image = runBlocking {
+      imageService.getLatestNamedImageWithAllRegionsForAppVersion(
+        appVersion = AppVersion.parseName(version.replace("~", "_")),
+        account = "test", // TODO: update when we have "default image account" configuration
+        regions = constraint.regions.toList())
+    }
+      ?.imageName ?: error("Image not found for $version in all requested regions ($regions)")
+
+    val regionalJobs = runBlocking {
+      val source = getSourceServerGroups(deliveryConfig.application, constraint)
+      require(regions.all { source.containsKey(it) }) {
+        "Source cluster ${constraint.source.cluster} not available in all canary regions"
+      }
+
+      regions.associateWith { region ->
+        val sourceServerGroup = source.getValue(region)
+        constraint.toStageBase(
+          cloudDriverCache = cloudDriverCache,
+          storageAccount = constraint.storageAccount ?: DEFAULT_KAYENTA_STORAGE_ACCOUNT,
+          app = deliveryConfig.application,
+          control = sourceServerGroup.toKayentaStageServerGroup(constraint.capacity, "baseline", image),
+          experiment = sourceServerGroup.toKayentaStageServerGroup(constraint.capacity, "canary", image)
+        )
+      }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    return regionalJobs.map { (region, task) ->
+      val description = "Canary $version for ${deliveryConfig.application}/environment " +
+        "${targetEnvironment.name} in $region"
+
+      region to
+        try {
+          runBlocking {
+            taskLauncher.submitJobToOrca(
```

Comment:
I suspect because naming things and consistency are both difficult :) I think "task" is a better fit based on pre-keel naming conventions (i.e. keel calls the /task REST endpoint in orca, Tasks tab in deck). "Job" is used within orca in multiple places and multiple meanings, but not for an orca orchestration. I get tripped up by the juxtaposition of the `TaskLauncher` class containing "task" but its only methods use "job". I'm into the idea of refactoring for naming consistency.